tcmur_device.c

```diff
@@ -238,33 +238,99 @@ int tcmu_cancel_lock_thread(struct tcmu_device *dev)
 	return ret;
 }
 
-/*
- * The initiator could have done FO/FB while no IO was
- * in flight so we did not get notified about losing the
- * lock. Update lock state now to avoid firing the error
- * handler later.
- */
-int tcmu_update_dev_lock_state(struct tcmu_device *dev)
+void tcmu_release_dev_lock(struct tcmu_device *dev)
 {
 	struct tcmur_handler *rhandler = tcmu_get_runner_handler(dev);
 	struct tcmur_device *rdev = tcmu_get_daemon_dev_private(dev);
-	int state;
+	int new_state = TCMUR_DEV_LOCK_LOCKED;
 
-	if (!rhandler->get_lock_state)
-		return -1;
-
-	state = rhandler->get_lock_state(dev);
 	pthread_mutex_lock(&rdev->state_lock);
-	if (rdev->lock_state == TCMUR_DEV_LOCK_LOCKED &&
-	    state != TCMUR_DEV_LOCK_LOCKED) {
-		tcmu_dev_dbg(dev, "Updated out of sync lock state.\n");
-		rdev->lock_state = TCMUR_DEV_LOCK_UNLOCKED;
+	if (rdev->lock_state != TCMUR_DEV_LOCK_LOCKED) {
+		pthread_mutex_unlock(&rdev->state_lock);
+		return;
 	}
 	pthread_mutex_unlock(&rdev->state_lock);
-	return state;
+
+	new_state = rhandler->unlock(dev);
+	if (new_state != TCMUR_DEV_LOCK_UNLOCKED)
+		tcmu_dev_warn(dev, "Lock not cleanly released. State %d.\n",
+			      new_state);
+	/*
+	 * If we don't have a clean unlock we still report success and set
+	 * to unlocked to prevent new IO from executing in case the lock
+	 * is in a state where it cannot be fenced.
+	 */
+	pthread_mutex_lock(&rdev->state_lock);
+	rdev->lock_state = TCMUR_DEV_LOCK_UNLOCKED;
+	pthread_mutex_unlock(&rdev->state_lock);
+}
+
+int tcmu_get_lock_tag(struct tcmu_device *dev, uint16_t *tag)
+{
+	struct tcmur_handler *rhandler = tcmu_get_runner_handler(dev);
+	struct tcmur_device *rdev = tcmu_get_daemon_dev_private(dev);
+	int retry = 0, ret;
+
+	if (rdev->failover_type != TMCUR_DEV_FAILOVER_EXPLICIT)
+		return 0;
+
+	ret = rhandler->get_lock_tag(dev, tag);
+	tcmu_dev_dbg(dev, "Got rc %d tag %hu\n", ret, *tag);
+retry:
+	switch (ret) {
+	case 0:
+		break;
+	case -ENOENT:
+		/* No lock holder yet */
+		break;
+	case -ESHUTDOWN:
```

Comment:
Nit: seems to be leaking RBD-specific error codes back to the generic handler. Could you just re-use the existing lock state enums (a la ```TCMUR_DEV_LOCK_FENCED```)?

Reply:
I am going to use the TCMUR_DEV_LOCK codes like suggested above.

Note off list I said I was going to use the TCMU error codes being added in 403,  but that is not done yet. Will convert to them later.

Reply:
I flip flopped again on this. In this PR I left the -Exyz codes. In this patchset

https://github.com/open-iscsi/tcmu-runner/pull/413

I went and made all the core and callout IO path code including above use common TCMU_ status codes.